Validate LLM output #432

Adem-Bouatay opened the pull request:
Used JSON-Schema to validate the LLM's output.
the same thing can be done with YAML with yamale library

adeprez commented on lavague-core/lavague/core/extractors.py:
```diff
@@ -16,19 +17,32 @@ class JsonFromMarkdownExtractor(BaseExtractor):
-
-    def extract(self, markdown_text: str) -> str:
+   
+    def extract(self, markdown_text: str, action_shape_validator = None) -> str:
```
Since the extractor can be used with structures beyond just actions, I suggest naming the validator `shape_validator`

adeprez commented on lavague-core/lavague/core/extractors.py:
```diff
@@ -16,19 +17,32 @@ class JsonFromMarkdownExtractor(BaseExtractor):
+                #checks if the json returned from the llm matchs the schema
+                validate(instance=json.loads(match.group(1).strip()), schema=action_shape_validator)
+            except json.JSONDecodeError as e:
+                raise(f"Invalid JSON format: {e}")
+            except ValidationError as e:
+                raise(f"JSON does not match schema: {e}")
```
Can you raise a custom exception? Something like `ExtractionError`. It will make easier to catch it later

adeprez commented on lavague-core/lavague/core/navigation.py:
```diff
@@ -271,7 +294,11 @@ def execute_instruction_gradio(self, instruction: str, action_engine: Any):
+                    action = self.extractor.extract(response,action_shape_validator=self.action_shape_validator)
+                except Exception as e:
+                    logging_print.error(f"Error extracting action: {e}")
+                    break
```
Instead of `break` we can continue so the LLM gets another chance to generate valid action until we reach `n_attempts` limit

```suggestion
                    continue
```

adeprez commented on lavague-core/lavague/core/navigation.py:
```diff
@@ -465,7 +492,11 @@ def execute_instruction(self, instruction: str) -> ActionResult:
+                    action = self.extractor.extract(response,action_shape_validator=self.action_shape_validator)
+                except Exception as e:
+                    logging_print.error(f"Error extracting action: {e}")
+                    break
```
```suggestion
                    continue
```

adeprez reviewed (changes requested):
Thank you for your work on this MR. It looks great! I made a few suggestions.

Could you please update your branch with the latest changes from the main branch? There's a small import conflict in `extractors.py`, and a larger one in `navigation.py` due to indentation - for this one I would recommand to accept changes and then add the validator again.

Adem-Bouatay commented:
> Thank you for your work on this MR. It looks great! I made a few suggestions.
> 
> Could you please update your branch with the latest changes from the main branch? There's a small import conflict in `extractors.py`, and a larger one in `navigation.py` due to indentation - for this one I would recommand to accept changes and then add the validator again.

Hi, thank you for the review. I solved the conflicts and added your suggestions, one thing i removed the json validation from the navigation since you moved to YAML, i'll look for how to implement the same thing  with YAML

adeprez commented on lavague-core/lavague/core/navigation.py:
```diff
@@ -379,11 +400,12 @@ def execute_instruction_gradio(self, instruction: str, action_engine: Any):
         navigation_log_total.append(navigation_log)
 
         if not success:
-            self.history[-1] = ChatMessage(
-                role="assistant",
-                content=f"Instruction: {action_engine.curr_instruction}",
-                metadata={"title": f"❌ Step {action_engine.curr_step + 1}"},
+            self.history[-1] = (
+                self.history[-1][0],
+                f"❌ Step {action_engine.curr_step + 1}:\n{action_engine.curr_instruction}",
             )
+            self.history.append((None, None))
+
```
Same as conflict from lines -320 to +348

adeprez commented on lavague-core/lavague/core/navigation.py:
```diff
@@ -428,15 +451,16 @@ def execute_instruction(self, instruction: str) -> ActionResult:
 
-        rephrased_query = self.rephraser.rephrase_query(instruction)
-
+        action = self.rephraser.rephrase_query(instruction)
```
This part could remain untouched - rephraser will be removed soon so it will be easier to merge if there's minimal impact on it

adeprez reviewed (commented):
That you for the updates, I think there was some issues while merging? I see some additional changes that are related to work recently merged on the main branch

Adem-Bouatay commented:
> That you for the updates, I think there was some issues while merging? I see some additional changes that are related to work recently merged on the main branch

Thank you again. It looks like when i merged with main branch the navigation.py did not sync., the conflict should be solved now if not i'll refork and add the changes again

adeprez commented:
Thank you! Conflicts are resolved and this branch is merged
